### ISSUE TITLE
How to access elements in MongoDB\Model\BSONDocument

### ISSUE THREAD

**Issue by raeesiqbal** [mongodb/mongo-php-library#168]
```php
$Existing = $DB->selectCollection('domains')->findOne(['DomainName' => $BaseDomain]);
```
`$Existing` has a several elements inside, for example, `DomainName` or its `_id`. How do we access them?
[HHVM Hack Type Checker](https://docs.hhvm.com/hack/typechecker/introduction) reports that the `$Existing` is an object of type `MongoDB\Model\BSONDocument` and is not an object of type `KeyedContainer`, so, we cannot do a `$Existing['_id']`.
Now, how do we access the elements inside?

**Comment by moderndeveloperllc** [comment]
It's an object, so you can do `$Existing->_id`. If you have an existing PHP class that implements [MongoDB\BSON\Unserializable](http://php.net/manual/en/class.mongodb-bson-unserializable.php) you can unserialize the return directly to your own class. Check out the documentation here for more: http://mongodb.github.io/mongo-php-library/tutorial/bson/#type-maps

**Comment by raeesiqbal** [comment]
I'm using this in HackLang actually and `$Existing->_id` will not work because you can't apply static typing for each `MongoDB\Model\BSONDocument ` as they can be different all the time. But I just found out that you can do a
```
$Existing = $Existing->toArray();
```
Thanks for the time

**Comment by moderndeveloperllc** [comment]
@raeesiqbal Glad to try to help! I would really look into http://mongodb.github.io/mongo-php-library/tutorial/bson/#persistable-classes if you can. It's _very_ nice to have results just populate into a model class. It's like having a lite ODM.

**Comment by raeesiqbal** [comment]
I will look into it. Thanks

**Comment by jmikola** [comment]
@raeesiqbal: I've not worked much with HackLang's Collection interfaces, but the library's BSONDocument class, which we use by default, extends PHP's [ArrayObject](http://php.net/arrayobject) class, which implements [ArrayAccess](http://php.net/manual/en/class.arrayaccess.php). Does HackLang not allow you to use the offset accessors (e.g. `[]`) on ArrayObjects in [strict mode](https://docs.hhvm.com/hack/typechecker/modes#strict-mode)?

**Comment by raeesiqbal** [comment]
If you disable the type checker `$restult['_id']` will work just fine (maybe because of PHP compatibility). But the type checker reports that MongoDB\Model\BSONDocument must implement `[KeyedContainer](https://docs.hhvm.com/hack/reference/interface/HH.KeyedContainer/)` to be accessible in `['_id']` style

**Comment by raeesiqbal** [comment]
I tried to add `KeyedContainer` manually to `MongoDB\Model\BSONDocument`'s definition:
```php
class MongoDB\Model\BSONDocument extends ArrayObject implements MongoDB\BSON\Serializable, MongoDB\BSON\Unserializable, MongoDB\BSON\Type, KeyedContainer {
    public static function __set_state(array $properties);
    public function bsonSerialize();
    public function bsonUnserialize(array $data);
}
```
But it gives an error that `KeyedContainer` interface expects 2 arguments. Here is the definition for `KeyedContainer`:
```php
interface KeyedContainer<+Tk, +Tv> extends Container<Tv>, KeyedTraversable<Tk, Tv> {}
```

**Comment by jmikola** [comment]
I don't think BSONDocument should be implementing [KeyedContainer](https://docs.hhvm.com/hack/reference/interface/HH.KeyedContainer/), as that interface entails declaring a type for all keys and values. The values in your BSON document are going to vary based on the fields in your document schema.

I assume that since the library is written in PHP, any HackLang code interacting with it will need to use [partial mode](https://docs.hhvm.com/hack/typechecker/modes#partial-mode). I don't believe you mentioned what mode you were using, but I assumed it was strict.

I also dug up https://github.com/facebook/hhvm/issues/6745#issuecomment-171022143 from a few months ago, where @simonwelsh reported that "Hack doesn't support ArrayAccess on user land classes." It doesn't look like that note has been added to the [unsupported features documentation](https://docs.hhvm.com/hack/unsupported/introduction) yet.

**Comment by lindelius** [comment]
If possible, specify the typeMap-option to return the result as an array instead of an object. It's both faster (not by much, but it's faster) and you can access the fields as a normal array.

```
$driverOptions= [
    'typeMap' => [
        'root' => 'array', 
        'document' => 'array', 
        'array' => 'array'
    ]
];

$client = new \MongoDB\Client($uri, $uriOptions, $driverOptions);
```

**Comment by raeesiqbal** [comment]
@jmikola , I am using strict mode for all of my application except for the libraries written in PHP, like `mongo-php-library`. And I added `nullable array` as return type for `MongoDB\Collection`'s functions I'm using, for example:
```php
class MongoDB\Collection implements Stringish {
  public function find($filter = array (), array $options = array ()): Vector<array>;
  public function findOne($filter = array (), array $options = array ()): ?array;
}
```

And @lindelius HHVM actually does not complain in the runtime when I handle BSONDocument like array. And the type checker wont know if I have specified the `typeMap` and the result will be in array, because the `typeMap` is an array and type checker is only smart with shapes in this matter. The problem was with the type checker only which I kind of hacked. PHP and HackLang are pretty different languages, so, unless this library is re-written in HackLang, I think this is the best we can get.
